plusdeck/cli.py

```diff
@@ -0,0 +1,371 @@
+import asyncio
+from dataclasses import asdict, dataclass, is_dataclass
+from enum import Enum
+import functools
+import json
+import logging
+import os
+import sys
+from typing import Any, Callable, Coroutine, List, Literal, Optional
+import warnings
+
+try:
+    from typing import Self
+except ImportError:
+    Self = Any
+
+import click
+from serial.serialutil import SerialException
+
+from plusdeck.client import Client, create_connection, State
+from plusdeck.config import Config, GLOBAL_FILE
+
+logger = logging.getLogger(__name__)
+
+OutputMode = Literal["text"] | Literal["json"]
+
+
+@dataclass
+class Obj:
+    """
+    The main click context object. Contains options collated from parameters and the
+    loaded config file.
+    """
+
+    config: Config
+    global_: bool
+    port: str
+    output: OutputMode
+    timeout: Optional[float]
+
+
+LogLevel = (
+    Literal["DEBUG"]
+    | Literal["INFO"]
+    | Literal["WARNING"]
+    | Literal["ERROR"]
+    | Literal["CRITICAL"]
+)
+
+STATES: List[str] = [state.name for state in State]
+
+
+class PlusdeckState(click.Choice):
+    """
+    A Plus Deck 2C state.
+    """
+
+    name = "state"
+
+    def __init__(self: Self) -> None:
+        super().__init__(STATES)
+
+    def convert(
+        self: Self,
+        value: str,
+        param: Optional[click.Parameter],
+        ctx: Optional[click.Context],
+    ) -> State:
+        choice = super().convert(value, param, ctx)
+
+        return State[choice]
+
+
+STATE = PlusdeckState()
+
+
+def as_json(obj: Any) -> Any:
```

Comment:
This functionality is super overkill for the plusdeck client, since the only thing that gets logged is a string state. Does it make sense to remove these abstractions and implement this switching in the subscribe command? Or simply not support an `--output` flag?

Reply:
Given that I haven't implemented config management yet, it may still make sense to have these abstractions.

Reply:
Yes, now that config commands are in there and I have a need to log the config (a dataclass), these helpers make sense.